File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    static int N, ret = Integer.MAX_VALUE;
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static List<Ingredient> ingredients = new ArrayList<>();
    static boolean first = true;
    public static void main(String[] args) throws IOException {
        Init();
        solve();
    }

    public static void Init() throws IOException {
        N = Integer.parseInt(br.readLine());
        for(int i = 0; i < N; i++){
            st = new StringTokenizer(br.readLine(), " ");
            ingredients.add(new Ingredient(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())));
        }
    }

    public static void go(int idx, int S, int B){
        if(!first)
            ret = Math.min(ret, Math.abs(S- B));

        for(int i = idx; i < N; i++){
            first = false;
            go(i + 1, S * ingredients.get(i).S, B + ingredients.get(i).B);
        }
    }

    public static void solve(){
        go(0,1,0);
        System.out.print(ret);
    }
}

class Ingredient{
    int S;
    int B;

    public Ingredient(int s, int b) {
        S = s;
        B = b;
    }
}

File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>
#include <climits>

using namespace std;

int N;
int ret = INT_MAX;
vector<pair<int, int>> ingredient;
bool first = true;

void FastIO() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
}

void Init() {
    cin >> N;
    for (int i = 0; i < N; i++) {
        int a, b;
        cin >> a >> b;
        ingredient.push_back({a, b});
    }
}

void dfs(int idx, int S, int B) {
    if (!first)
        ret = min(ret, abs(S - B));

    for (int i = idx; i < N; i++) {
        first = false;
        dfs(i + 1, S * ingredient[i].first, B + ingredient[i].second);
    }
}

void solve() {
    dfs(0, 1, 0);
    cout << ret;
}

int main(void) {
    FastIO();
    Init();
    solve();
    return 0;
}